### PR TITLE
feat(frontend): report live session activity on the workspace rail tiles

### PR DESCRIPTION
**Pull request by Felixmil**
## Summary

- A user can keep several workspaces open, but the left rail shows only a workspace icon and which workspace is selected. To learn that a session in another workspace finished its turn, needs an answer, or failed, the user has to switch to that workspace. Each tile now reports the live activity of its own workspace, so the rail can be scanned without leaving the current workspace.
- Out of scope: counts or per-session detail in a badge, badges as controls, indicators anywhere else (repository selector, workspace settings), notification delivery and wording, and how a session reaches a state. This change only reports states that already exist.

## Reviewer context

The frontend's session state is scoped to the selected workspace. `createAgentSessionsStore` is reset on every workspace switch and `useShellAgentActivity` returns an empty summary for a non-active repository path, so no existing read model can answer "what is happening in workspace B" while workspace A is selected. The live data does exist per repository: `observeAgentSessionLive({ repoPath }, listener)` already runs for every workspace in the rail for notifications, and both shells support several subscribers on one repository path.

This PR therefore adds one reduced live projection per workspace that produces three booleans, instead of growing a host contract. Two alternatives were rejected. Aggregating in the host would need new persisted state and a new event, because the host snapshot has no error state; a failed session is a renderer-side conclusion drawn from `session_error` transcript events. Reusing the notification workspace observer would tie badges to notification baselines, which is a non-goal.

## Design and tradeoffs

- Status derivation is shared, not restated. `projectObservedSessionActivity` and `projectSessionTranscriptActivity` are exported from the Agent Studio read model with widened parameter types, so the tile and the workspace inside cannot drift apart. These are type and visibility changes only; Agent Studio behavior is unchanged.
- Live state stays outside TanStack Query. The only query read is the shared archived chat list, used to exclude archived sessions.
- Failures are reported, not hidden. A `fault`, a `transcript_gap`, a rejected observation start, or a failed archived-list read puts the tile into the unavailable state, which shows the red badge and carries the reported reason in its hover and assistive text. There is no retry, no second data path, and no fallback to older data.
- The rail counts every non-subagent live session reported for the workspace. Applying the Agent Studio admission rule would mean loading every task and every task session record for every workspace, a large read amplification for three booleans. A session the runtime reports but OpenDucktor did not register normally reads as idle and produces no badge, so the failure direction reports real runtime activity rather than hiding it.
- Badge appearance changed during review. The badges started as colored glyphs copied from the Agent Studio task tabs, but they read badly in the dark theme, so the final form is a plain colored dot per state, using the same tokens as the session status dots of the workspace sessions page. This drops the spec's "not color alone" property; the state name is still carried in full by the hover text and the assistive text of each badge.

## Preview

Preview of the 3 status badges:

https://github.com/user-attachments/assets/efcb6f4b-c817-4781-a698-d67841059325

## Testing

- [x] `bun run lint`
- [x] `bun run typecheck`
- [x] `bun run test`
- [x] `bun run build`
- [ ] Not applicable

All four were run from the repository root after the rebase onto `main`, along with `bun run format:check`. The frontend suite reports 5893 pass and 0 fail.

## Checklist

- [x] I updated docs when behavior, workflows, runtime contracts, or setup changed
- [x] I added or updated relevant tests
- [x] I kept failures explicit instead of adding fallback logic
- [x] I linked related issues or context below

`docs/agent-orchestrator-module-map.md` now scopes the live-state rule to the Agent Studio store and names the shared status rules. New tests cover the projection envelope handling, the badge fold, the observer lifecycle, and the rail rendering, including badge order, the unavailable reason, that a click on a badge still selects the workspace, and that the drag overlay copy carries its own badges.

## Related Issues

No GitHub issue. Tracked as OpenDucktor task `openduckto-p3y2`.

## Breaking Changes

- [x] No breaking changes
- [ ] Breaking changes are described below

## Additional Context

The badges sit in a strip absolutely positioned inside the tile button, so tile size, rail width, and tile spacing are unchanged, clicks and drag still reach the button, and the drag overlay copy shows the same badges. The active dot keeps a soft opacity pulse and is listed in the `prefers-reduced-motion` block.

PR #926 also touches the workspace rail tiles, for the workspace abbreviation and tile color. Whichever lands second may need a small merge in `workspace-rail.tsx`.

## Summary by CodeRabbit

- **New Features**
  - Workspace tiles now display activity badges for active sessions, errors, and required user input.
  - Activity indicators stay synchronized with Agent Studio session status across workspaces.
  - Unavailable activity displays a reason when live updates or session records cannot be retrieved.
  - Activity indicators update as session records load, fail, or recover.
  - Badges include accessible labels and descriptions for assistive technologies.
- **Style**
  - Running activity indicators now pulse, with animation disabled for reduced-motion preferences.
- **Documentation**
  - Clarified consistent live activity projection across the workspace rail and Agent Studio.

### CODE REVIEW

File: packages/frontend/src/features/workspace-activity/workspace-activity-state.ts [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+ * same attribution the Agent Studio read model performs.
+ */
+export const foldWorkspaceActivityBadges = (
+  sessions: ReadonlyMap<string, WorkspaceActivitySession>,
+  archivedSessionKeys: ReadonlySet<string>,
+): WorkspaceActivityBadges => {
+  const counted = new Map<string, WorkspaceActivitySession>();
+  for (const [key, session] of sessions) {
+    if (!archivedSessionKeys.has(key)) {
+      counted.set(key, session);
```
[the remark this answers is not in the document]
Comment (Felixmil) [reply packages/frontend/src/features/workspace-activity/workspace-activity-state.ts]: Fixed in ee833c248.

`foldWorkspaceActivityBadges` now resolves ancestry against the full session map before counting. A session is excluded when its own key or any reachable ancestor key is archived, so a live subagent of an archived chat no longer becomes an orphan owner and produces a badge. The cycle guard of the existing walk is kept, and a session whose parent is not reported still counts, which is the deliberate rule for unreported parents.

Regression test added in `workspace-activity-state.test.ts`: an archived parent with a running child folds to no badge.

File: packages/frontend/src/features/workspace-activity/workspace-activity-observer.ts [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+        }
+        observation.archivedLoaded = true;
+        emit();
+      })
+      .catch((cause: unknown) => {
+        if (observation.cancelled) {
+          return;
+        }
+        observation.archivedError = errorMessage(cause);
+        emit();
```
[the remark this answers is not in the document]
Comment (Felixmil) [reply packages/frontend/src/features/workspace-activity/workspace-activity-observer.ts]: Correct, and fixed in ee833c248.

The observer no longer keeps the failed archived read on the observation. The archived chat state is read from the source on every fold, as `{ status: "unknown" | "ready" | "error" }`, so a read that fails and later succeeds clears the tile by itself. The provider derives that state from the query state of the shared archived chat list, and its cache subscription now notifies on failed record queries as well as successful ones, so the tile also reports the failure when it happens.

No retry and no second data path were added: the tile reports what the record cache currently says.

Regression test added in `workspace-activity-observer.test.ts`: a failed archived read reports unavailable, then reports ready once the read recovers.

File: packages/frontend/src/components/layout/workspace-rail-activity-badges.tsx [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+ * Badge fill, matching the session status dots of the workspace sessions page.
+ *
+ * The badges carry no glyph, so the tile states are told apart by color, and
+ * the hover text and assistive text of each badge name the state in full.
+ */
+const WORKSPACE_ACTIVITY_BADGE_COLORS = {
+  inputRequired: "bg-warning-accent",
+  error: "bg-destructive",
+  active: "workspace-rail-status-running-dot bg-status-running",
+} satisfies Record<WorkspaceActivityBadgeKey, string>;
```
[the remark this answers is not in the document]
Comment (Felixmil) [reply packages/frontend/src/components/layout/workspace-rail-activity-badges.tsx]: Not changed, this one is deliberate.

The badges started as distinct glyphs, exactly for the reason you give: `CircleAlert` for input required, `TriangleAlert` for error, and a running glyph for active. They read badly in the dark theme at tile-badge size, so the decision in review was to drop the glyphs and use the plain colored dots the workspace sessions page already uses for session status. The tile indicator now matches the indicator the same session shows inside the workspace, which was a requirement of this task.

The accepted cost is that color is the only visual difference between the three states. What remains for a user who cannot tell the colors apart: each badge carries a `title` and screen reader text naming the state in full ("Sessions waiting for input", "Sessions failed", "Sessions running"), and the strip is associated with the tile through `aria-describedby`.

Leaving the thread open so the maintainer can weigh in on whether the glyphs should come back in another form.

File: packages/frontend/src/state/providers/workspace-activity-provider.tsx [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+      read: (workspaceId) => {
+        const state = queryClient.getQueryState<WorkspaceSession[]>(
+          workspaceSessionQueryKeys.list(workspaceId, true),
+        );
+        if (state?.data) {
+          return { status: "ready", keys: toArchivedSessionKeys(state.data) };
+        }
+        if (state?.status === "error") {
+          return { status: "error", reason: errorMessage(state.error) };
+        }
```
[the remark this answers is not in the document]
Comment (Felixmil) [reply packages/frontend/src/state/providers/workspace-activity-provider.tsx]: Valid, and fixed in e2eefa7c7. The error check now runs before the data check.

You are right about the cache behavior: a failed refetch keeps the data of the last success, so reading the data first answered from data the source had disowned. That is also the one thing the spec for this change forbids, a fallback to older data to fill the gap, so the tile now reports the failure.

The port moved out of the provider into `state/workspace-activity/archived-sessions-port.ts`, so the read order is covered by a test instead of only by review: `archived-sessions-port.test.ts` seeds a successful list, fails the next read, and asserts the port reports the error rather than the stale keys. I checked that the test fails against the previous order.

File: packages/frontend/src/features/workspace-activity/workspace-activity-state.ts [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+  key: string,
+): boolean => {
+  let current = key;
+  const visited = new Set([key]);
+  for (;;) {
+    if (archivedSessionKeys.has(current)) {
+      return true;
+    }
+    const parentKey = sessions.get(current)?.parentKey ?? null;
+    if (parentKey === null || visited.has(parentKey) || !sessions.has(parentKey)) {
```
[the remark this answers is not in the document]
Comment (Felixmil) [reply packages/frontend/src/features/workspace-activity/workspace-activity-state.ts]: Valid, and fixed in 06d858414.

The walk no longer stops at a parent that the live stream does not report. The archived keys come from the chat records, not from the live stream, so an ancestor key is now checked against them even when that ancestor is absent from the session map. The cycle guard is unchanged, and the next step ends the walk by itself because a missing session has no parent key.

This is more than an edge case: archiving a chat stops its session, so the archived parent of a lingering subagent is often exactly the parent that is no longer reported.

I added the regression test first and watched it fail against the previous walk: `workspace-activity-state.test.ts` now folds a running subagent whose archived parent is absent from the live sessions and expects no badge.

**Comment by Maxsky5** [comment]
Hi @Felixmil 👋 

That's again a great idea.

It's been a while that I'm trying to find a way to have a global view of the running sessions and pending actions across workspaces. Still didn't find a design I'm happy with.

Until then, the badges in the workspace rails will be very useful 🙂

I will just push a commit to align the running session badge with the glowing dot we have in the agent studio tabs, and rebase on main.

Thanks again for your contribution!
